src/ray/object_manager/object_buffer_pool.cc

```diff
@@ -0,0 +1,235 @@
+    switch (pair.second.state_type) {
+    case BufferStateType::GET: {
+      AbortGet(pair.first);
+    } break;
+    case BufferStateType::CREATE: {
+      AbortCreate(pair.first);
+    } break;
+    }
+  }
+  RAY_CHECK(chunk_info_.empty());
+  RAY_CHECK(buffer_state_.empty());
+  RAY_CHECK(create_failure_buffers_.empty());
+  // Disconnect plasma clients.
+  for (const auto &client : clients) {
+    ARROW_CHECK_OK(client->Disconnect());
+  }
+  available_clients.clear();
+  clients.clear();
+}
+
+ray::Status ObjectBufferPool::SealCreate(const ObjectID &object_id) {
+  const plasma::ObjectID plasma_id = ObjectID(object_id).to_plasma_id();
+  chunk_info_.erase(object_id);
+  std::shared_ptr<plasma::PlasmaClient> store_client = buffer_state_[object_id].client;
+  if (create_failure_buffers_.count(object_id) == 0) {
+    // This is a successful create.
+    ARROW_CHECK_OK(store_client->Seal(plasma_id));
+    ARROW_CHECK_OK(store_client->Release(plasma_id));
+  } else {
+    // This is a failed create due to failure to allocate buffer.
+    create_failure_buffers_.erase(object_id);
+    RAY_LOG(ERROR) << "Receive Failed";
+  }
+  ReleaseObjectStore(store_client);
+  buffer_state_.erase(object_id);
+  return ray::Status::OK();
+}
+
+ray::Status ObjectBufferPool::AbortCreate(const ObjectID &object_id) {
+  const plasma::ObjectID plasma_id = ObjectID(object_id).to_plasma_id();
+  chunk_info_.erase(object_id);
+  std::shared_ptr<plasma::PlasmaClient> store_client = buffer_state_[object_id].client;
+  if (create_failure_buffers_.count(object_id) == 0) {
+    // This is a failed create due to error on receiving data.
+    ARROW_CHECK_OK(store_client->Release(plasma_id));
+    ARROW_CHECK_OK(store_client->Abort(plasma_id));
+  } else {
+    // This is a failed create due to failure to allocate buffer.
+    create_failure_buffers_.erase(object_id);
+    RAY_LOG(ERROR) << "Receive Failed";
+  }
+  ReleaseObjectStore(store_client);
+  buffer_state_.erase(object_id);
+  return ray::Status::OK();
+}
+
+uint64_t ObjectBufferPool::BuildChunks(const ObjectID &object_id, uint8_t *data,
+                                       uint64_t data_size, uint64_t metadata_size) {
+  uint64_t space_remaining = data_size;
+  std::vector<ChunkInfo> chunks;
+  int64_t position = 0;
+  while (space_remaining) {
+    position = data_size - space_remaining;
+    if (space_remaining < chunk_size_) {
+      chunks.emplace_back(data + position, space_remaining, data_size, metadata_size,
+                          ray::Status::OK());
+      space_remaining = 0;
+    } else {
+      chunks.emplace_back(data + position, chunk_size_, data_size, metadata_size,
+                          ray::Status::OK());
+      space_remaining -= chunk_size_;
+    }
+  }
+  chunk_info_[object_id] = chunks;
+  return chunk_info_[object_id].size();
+}
+
+std::shared_ptr<plasma::PlasmaClient> ObjectBufferPool::GetObjectStore() {
+  if (available_clients.empty()) {
```

Comment:
Now that the pool is at the granularity of object buffers instead of plasma clients, can we have a single client that is shared by all of the buffers?